[FIRRTL] Document baseline naming and what symbols imply (#3197)
llvm/circt: docs/RationaleFIRRTL.md

> ## Naming

Reviewer: I'd move the definition of _named_ earlier in the text. A lot of the text hinges on understanding what this means. It may be good to also contrast this with LLVM/MLIR where naming is more explicit, e.g., `%0` vs. `%foo`.

> Names in verilog form part of the public-api of a design.  It is therefore

Reviewer 2: ```suggestion
Names in verilog form part of the public API of a design.  It is therefore
```

> critical that Chisel, and by extension FIRRTL, have a language-level semantic 

Reviewer 2: ```suggestion
critical that Chisel, and by extension FIRRTL, have language-level semantics
```

about how entities are named and how named entities are used and transformed.  
This must specify which entities with names in Chisel generate predictable 

> output.  Since names server multiple purposes in a design; for example, 

Reviewer 2: ```suggestion
output.  Since names serve multiple purposes in a design; for example, 
```

> debugging, test-bench attachment, hooks for physical layout, etc; we must 

Reviewer 2: Not sure about use of semicolons for bracketing what's usually a parenthetical (or with some sort of dashes around).  But works for me.

Reviewer: Agree, I'd just use `(e.g., a, b, or c)`.

balance multiple needs.  This section describes the base semantics, which are 
conservative and aimed at enabling debugging.  _It is expected that the 

> compiler provide flags to relax these rules to produce more synthesis or 

Reviewer 2: ```suggestion
compiler provide flags to relax these rules to produce more synthesis- or 
```
(we don't want to produce more synthesis, but synthesis-ready output)

production-ready output_.

Modules shall use the name given in Chisel, unless they conflict with a verilog 

> reserved, not withstanding de-duplication or relevant annotations on the module.

Reviewer 2: ```suggestion
reserved word, not withstanding de-duplication or relevant annotations on the module.
```
(or some noun after 'reserved'?)

Instances shall use the name given in Chisel, unless they conflict with a 

> verilog reserved.  Instances have preferential use of the name in the output in 

Reviewer 2: ```suggestion
verilog reserved word.  Instances have preferential use of the name in the output in 
```
(or appropriate noun, again)

case of a conflict, after ports.

> Chisel's provides a "Don't Touch" annotation to protect entities from 

Reviewer 2: ```suggestion
Chisel provides a "Don't Touch" annotation to protect entities from 
```

transformation.  A don't touch on a wire or node produces a wire in verilog and 
preserves the data-flow through that wire.  Even a wire driven by a constant 
shall not have the constant forwarded around the wire.  This is because a 
"don't touch" annotation signals the possible public use of a wire and one 
common use is to provide a place to drive a new value into the logic from an 
external test-bench.  If the node or wire is named (and it always should be for 
Chisel don't touch), this name is used, unless it conflicts with a verilog 
reserved word.  This wire has preferential use of the name in the output in 
case of a conflict, after ports and instances.

> Named wires and nodes in firrtl shall appear as a wire in the output verilog.  

Reviewer 2: ```suggestion
Named wires and nodes in FIRRTL shall appear as a wire in the output verilog.  
```

There is no requirement that data-flow through a wire be maintained, only that 
the data-flow into a wire be maintained.  This allows bypassing and forwarding 
around wires who exist solely because of their name.  An implementation may 
choose to not bypass trivial wires to reduce unused wire lint warnings, but 
shouldn't cause other lint warnings to avoid unused wire warnings.  A named 
wire without a symbol is thus equivalent to a named read-probe in the circuit.

Any name of an entity inside a module which starts with `_` may be discarded.  
This name pattern indicates the name is for convenience in the Chisel code 
(often temporaries are required) and there is no expectation it exist in the 
output.

### Mandatory Renaming

We want the naming of Verilog objects to match the names used in

> the original CHISEL, but in several passes, there is mandatory renaming.  It is 

Reviewer: ```suggestion
the original Chisel, but in several passes, there is mandatory renaming.  It is 
```

important that this be a predictable transformation.  For example, after 
bundles are replaced with scalars in the lower-types pass, each field should be 

> prefixed with the bundle name:

Reviewer: Eventually expanding on the _exact_ algorithm here would be good. Not blocking for this PR.

if the object has an array attribute `annotations` containing the attribute
`{class = "firrtl.transforms.DontTouchAnnotation}`.

## Symbols and Inner Symbols

Symbols and Inner Symbols are documented in the [symbol rationale](RationaleSymbols.md).
This section documents how symbols are used, their interaction with don't 
touch, and the semantics imposed by them.

> Public Symbols indicate there are used of an entity outside the analysis scope 

Reviewer 2: ```suggestion
Public Symbols indicate there are uses of an entity outside the analysis scope 
```

> of the compiler.  This requires the the entity be preserved in such a way as 

Reviewer 2: ```suggestion
of the compiler.  This requires the entity be preserved in such a way as 
```

the operations possible in the target language have the expected effect.  For 
example, a wire or port with a public symbol may be used by name in a test 

> bench to read or write new values into the circuit.  Therefor, these wires 

Reviewer 2: ```suggestion
bench to read or write new values into the circuit.  Therefore, these wires 
```

cannot be detached form their original dataflow as this would break the remote 
write case, nor can their input dataflow be changed, as this would break the 
remote read case.  They cannot be renamed, as this would break all remote access.

Private Symbols indicate there are symbolic references to the entity, but they 
are all within the scope of the compiler's IR and analysis.  An entity with a 
private symbol may be arbitrarily transformed, so long as the transformation is 
semantic preserving with respect to all uses of the private symbol.  If it can 

> be proved a wire with a private symbol is only read from via it's symbol and 

Reviewer 2: ```suggestion
be proved a wire with a private symbol is only read from via its symbol and 
```

not written to, for example, the input can forwarded to the output (bypassing 
the wire) safely.  If a private symbol is unused, it may be removed.  Private 
symbols impose no restriction on output, they only exist to enable non-local 
effects in the IR.

Don't touch is implemented as a public symbol on an entity.  A conservative 
interpretation of "don't touch", and a common use case, is that the entity is 
referred to by a testbench in unknown ways.  This implies no transformation 
which would change observed behavior if the entity was arbitrarily read or 
written to remotely.  This further implies the existence of the entity in the 
output.

> Importantly the existance of a symbol doesn't specify whether something is 

Reviewer 2: ```suggestion
Importantly the existence of a symbol doesn't specify whether something is 
```
